pkg/collectors/log_follower.go

```diff
@@ -0,0 +1,313 @@
+	if seen, ok := logs.seenLines[line.Timestamp]; ok {
+		for _, seenContent := range seen {
+			if seenContent == line.Content {
+				// Already have this line skip it
+				return
+			}
+		}
+	}
+	logs.seenLines[line.Timestamp] = append(logs.seenLines[line.Timestamp], line.Content)
+	logs.pruned = false
+	var err error
+	if logs.withTimeStamps {
+		_, err = writer.WriteString(line.Raw + "\n")
+	} else {
+		_, err = writer.WriteString(line.Content + "\n")
+	}
+	if err != nil {
+		log.Error(fmt.Errorf("failed to write log output to file"))
+	}
+}
+
+func (logs *LogsCollector) writeToLogFile() {
+	logs.wg.Add(1)
+	defer logs.wg.Done()
+
+	fileHandle, err := os.OpenFile(logs.logsOutputFileName, os.O_CREATE|os.O_WRONLY, logFilePermissions)
+	utils.IfErrorExitOrPanic(err)
+	defer fileHandle.Close()
+
+	for {
+		select {
+		case <-logs.quit:
+			// Consume the rest of the lines so we don't miss lines
+			for len(logs.lines) > 0 {
+				line := <-logs.lines
+				logs.consumeLine(line, fileHandle)
+			}
+			return
+		case line := <-logs.lines:
+			logs.consumeLine(line, fileHandle)
+		default:
+			if !logs.pruned {
+				log.Info("log: pruning seen lines")
+				keepTime := logs.lastPoll.Add(-2 * followDuration)
+				for key := range logs.seenLines {
+					if key.Sub(keepTime) < 0 {
+						delete(logs.seenLines, key)
+					}
+				}
+				logs.pruned = true
+			} else {
+				time.Sleep(time.Microsecond)
+			}
+		}
+	}
+}
+
+func processLine(line string) *ProcessedLine {
+	splitLine := strings.SplitN(line, " ", 2) //nolint:gomnd // moving this to a var would make the code less clear
+	if len(splitLine) < 2 {                   //nolint:gomnd // moving this to a var would make the code less clear
+		return nil
+	}
+	timestampPart := splitLine[0]
+	lineContent := splitLine[1]
+	timestamp, err := time.Parse(time.RFC3339, timestampPart)
+	if err != nil {
+		// This is not a value line something went wrong
+		return nil
+	}
+	return &ProcessedLine{
+		Timestamp: timestamp,
+		Content:   lineContent,
+		Raw:       line,
+	}
+}
+
+func (logs *LogsCollector) processLines(line string) (string, time.Time) {
+	var lastLine *ProcessedLine
+	if strings.ContainsRune(line, lineDelim) {
+		lines := strings.Split(line, "\n")
+		for n := 0; n < len(lines)-2; n++ {
+			processed := processLine(line)
+			log.Debug("logs: processed line: ", processed)
+			if processed != nil {
+				logs.lines <- processed
+				lastLine = processed
+			}
+		}
+		line = lines[len(lines)-1]
+	}
+	if lastLine == nil {
+		return line, time.Time{}
+	}
+	return line, lastLine.Timestamp
+}
+
+func durationPassed(first, current time.Time, duraration time.Duration) bool {
+	log.Info("logs: duration first: ", first)
+	log.Info("logs: duration current: ", current)
+	log.Info("logs: duration diff: ", first.Sub(current))
+	log.Info("logs: duration hads passed: ", !first.IsZero() && !current.IsZero() && duraration <= first.Sub(current))
```

Comment:
Do we need these logs?

Reply:
I've moved them to debug for the moment but I'll probably clean them up once I sort out this set of commits from the code review changes.